integration/linker/link-packages-test.js

```diff
@@ -0,0 +1,98 @@
+const {getNpmPackagesFromRunfiles} = require('../runfile-helpers');
+const {readFileSync} = require('fs');
+const {join} = require('path');
+const babel = require('@babel/core');
+const {default: traverse} = require('@babel/traverse');
+const glob = require('glob');
+const chalk = require('chalk');
+
+/** File system used by the Angular linker plugin. */
+const fileSystem = new NodeJSFileSystem();
+/** Logger used by the Angular linker plugin. */
+const logger = new ConsoleLogger(LogLevel.info);
+/** List of NPM packages available in the Bazel runfiles. */
+const npmPackages = getNpmPackagesFromRunfiles();
+/** Whether any package could not be linked successfully. */
+let failedPackages = false;
+
+// Iterate through all determined NPM packages and ensure that entry point
+// files can be processed successfully by the Angular linker.
+for (const pkg of npmPackages) {
+  const {failures, passedFiles} = testPackage(pkg);
+
+  console.info(chalk.cyan(`------- Package: @angular/${pkg.name} -------`));
+  console.info(`Passed files: ${passedFiles.length}`);
+  console.info(`Failed files: ${failures.length}`);
+
+  if (failures.length > 0) {
+    failures.forEach(({debugFileName, error}) => {
+      console.error(` •  ${chalk.yellow(debugFileName)}: ${error}`);
+    });
+    failedPackages = true;
+  }
+
+  console.info('-------------------------------------');
+  console.info();
+}
+
+if (failedPackages) {
+  console.error(chalk.red(`✘ Not all packages could be linked successfully. See errors above.`));
+  // If there are failures, exit the process with a non-zero exit code. Bazel
+  // uses exit code `3` to indicate non-fatal test failures.
+  process.exitCode = 3;
+} else {
+  console.info(chalk.green('✓ All packages have been successfully linked.'));
+}
+
+/**
+ * Tests the specified package against the Angular linker plugin.
+ * @param pkg Package being tested.
+ * @returns An object containing linker failures and passed files.
+ */
+function testPackage(pkg) {
+  const entryPointFesmFiles = glob.sync(`fesm2015/**/*.js`, {cwd: pkg.pkgPath});
+  const passedFiles = [];
+  const failures = [];
+
+  // Iterate through each entry point and confirm it's found Angular declarations from a partial
+  // compilation can be linked to the corresponding Angular definitions without errors.
+  for (const fesmFileName of entryPointFesmFiles) {
+    const diskFilePath = join(pkg.pkgPath, fesmFileName);
+    const debugFileName = join(pkg.name, fesmFileName);
+    const fileContent = readFileSync(diskFilePath, 'utf8');
+    const linkerPlugin = createEs2015LinkerPlugin({fileSystem, logger});
+
+    try {
+      const {ast} = babel.transformSync(fileContent, {
+        ast: true,
+        filename: diskFilePath,
+        filenameRelative: debugFileName,
+        plugins: [linkerPlugin]
+      });
+
+      // Naively check if there are any Angular declarations left that haven't been linked.
+      traverse(ast, {
```

Comment:
I think you can speed this up by using `traverse.cheap()`?

Reply:
Yeah, though I think that doesn't give us the `NodePath` and therefore makes it difficult to generate a nice error as done with `buildCodeFrameError`.

Reply:
Fair enough. It probably makes little difference to the overall speed either.